File: modules/stego-core/src/main/java/com/shterneregen/securelan/stego/service/impl/BmpSteganographyService.java
package com.shterneregen.securelan.stego.service.impl;

import com.shterneregen.securelan.crypto.model.PasswordEncryptedData;
import com.shterneregen.securelan.crypto.workflow.PasswordFileCryptoWorkflow;
import com.shterneregen.securelan.stego.model.BmpCapacity;
import com.shterneregen.securelan.stego.model.ExtractedStegoPayload;
import com.shterneregen.securelan.stego.model.StegoContentType;
import com.shterneregen.securelan.stego.service.SteganographyService;
import com.shterneregen.securelan.stego.service.impl.internal.BmpImage;
import com.shterneregen.securelan.stego.service.impl.internal.StegoHeader;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Objects;

public final class BmpSteganographyService implements SteganographyService {
    private final PasswordFileCryptoWorkflow passwordFileCryptoWorkflow;

    public BmpSteganographyService(PasswordFileCryptoWorkflow passwordFileCryptoWorkflow) {
        this.passwordFileCryptoWorkflow = Objects.requireNonNull(passwordFileCryptoWorkflow, "passwordFileCryptoWorkflow");
    }

    @Override
    public BmpCapacity inspect(byte[] bmpBytes) {
        BmpImage image = BmpImage.parse(bmpBytes);
        return new BmpCapacity(
                image.width(),
                image.height(),
                image.bitsPerPixel(),
                image.carrierByteCount(),
                StegoHeader.BYTE_LENGTH,
                image.maxPayloadBytes(StegoHeader.BYTE_LENGTH)
        );
    }

    @Override
    public byte[] hide(byte[] bmpBytes, byte[] payload, StegoContentType contentType) {
        Objects.requireNonNull(contentType, "contentType");
        return hideInternal(bmpBytes, payload, contentType, false);
    }

    @Override
    public ExtractedStegoPayload extract(byte[] bmpBytes) {
        BmpImage image = BmpImage.parse(bmpBytes);
        byte[] headerBytes = image.readLeastSignificantBits(bmpBytes, StegoHeader.BYTE_LENGTH);
        StegoHeader header = StegoHeader.read(headerBytes);
        int maxPayloadBytes = image.maxPayloadBytes(StegoHeader.BYTE_LENGTH);
        if (header.payloadLength() > maxPayloadBytes) {
            throw new IllegalArgumentException("Hidden payload length exceeds BMP capacity");
        }
        byte[] hiddenBytes = image.readLeastSignificantBits(bmpBytes, StegoHeader.BYTE_LENGTH + header.payloadLength());
        byte[] payload = new byte[header.payloadLength()];
        System.arraycopy(hiddenBytes, StegoHeader.BYTE_LENGTH, payload, 0, payload.length);
        return new ExtractedStegoPayload(header.contentType(), header.encrypted(), payload);
    }

    @Override
    public byte[] extractPayload(byte[] bmpBytes) {
        ExtractedStegoPayload extracted = extract(bmpBytes);
        requirePlain(extracted);
        return extracted.payload();
    }

    @Override
    public byte[] hideText(byte[] bmpBytes, String message) {
        Objects.requireNonNull(message, "message");
        return hide(bmpBytes, message.getBytes(StandardCharsets.UTF_8), StegoContentType.UTF8_TEXT);
    }

    @Override
    public String extractText(byte[] bmpBytes) {
        ExtractedStegoPayload extracted = extract(bmpBytes);
        requirePlain(extracted);
        return extracted.asUtf8String();
    }

    @Override
    public byte[] hideEncryptedPayload(byte[] bmpBytes, byte[] payload, char[] password) {
        return hideEncrypted(bmpBytes, payload, StegoContentType.BINARY, password);
    }

    @Override
    public byte[] extractEncryptedPayload(byte[] bmpBytes, char[] password) {
        return extractEncrypted(bmpBytes, password, StegoContentType.BINARY);
    }

    @Override
    public byte[] hideEncryptedText(byte[] bmpBytes, String message, char[] password) {
        Objects.requireNonNull(message, "message");
        return hideEncrypted(bmpBytes, message.getBytes(StandardCharsets.UTF_8), StegoContentType.UTF8_TEXT, password);
    }

    @Override
    public String extractEncryptedText(byte[] bmpBytes, char[] password) {
        byte[] decrypted = extractEncrypted(bmpBytes, password, StegoContentType.UTF8_TEXT);
        return new String(decrypted, StandardCharsets.UTF_8);
    }

    private byte[] hideEncrypted(byte[] bmpBytes, byte[] payload, StegoContentType contentType, char[] password) {
        Objects.requireNonNull(payload, "payload");
        requirePassword(password);
        PasswordEncryptedData encryptedData = passwordFileCryptoWorkflow.encrypt(payload, password);
        return hideInternal(bmpBytes, serialize(encryptedData), contentType, true);
    }

    private byte[] extractEncrypted(byte[] bmpBytes, char[] password, StegoContentType expectedContentType) {
        requirePassword(password);
        ExtractedStegoPayload extracted = extract(bmpBytes);
        if (!extracted.encrypted()) {
            throw new IllegalStateException("Extracted payload is not encrypted");
        }
        if (extracted.contentType() != expectedContentType) {
            throw new IllegalStateException("Unexpected encrypted payload type: " + extracted.contentType());
        }
        PasswordEncryptedData encryptedData = deserialize(extracted.payload());
        return passwordFileCryptoWorkflow.decrypt(encryptedData, password);
    }

    private byte[] hideInternal(byte[] bmpBytes, byte[] payload, StegoContentType contentType, boolean encrypted) {
        Objects.requireNonNull(payload, "payload");
        BmpImage image = BmpImage.parse(bmpBytes);
        int maxPayloadBytes = image.maxPayloadBytes(StegoHeader.BYTE_LENGTH);
        if (payload.length > maxPayloadBytes) {
            throw new IllegalArgumentException("Payload is too large for this BMP: "
                    + payload.length + " bytes requested, " + maxPayloadBytes + " bytes available");
        }
        StegoHeader header = new StegoHeader(contentType, encrypted, payload.length);
        byte[] hiddenBytes = ByteBuffer.allocate(StegoHeader.BYTE_LENGTH + payload.length)
                .put(header.write())
                .put(payload)
                .array();
        byte[] result = bmpBytes.clone();
        image.writeLeastSignificantBits(result, hiddenBytes);
        return result;
    }

    private void requirePlain(ExtractedStegoPayload extracted) {
        if (extracted.encrypted()) {
            throw new IllegalStateException("Extracted payload is encrypted; use encrypted extraction workflow");
        }
    }

    private void requirePassword(char[] password) {
        Objects.requireNonNull(password, "password");
        if (password.length == 0) {
            throw new IllegalArgumentException("password must not be empty");
        }
    }

    private byte[] serialize(PasswordEncryptedData encryptedData) {
        byte[] salt = encryptedData.salt();
        byte[] iv = encryptedData.iv();
        byte[] cipherText = encryptedData.cipherText();
        return ByteBuffer.allocate(Integer.BYTES * 3 + salt.length + iv.length + cipherText.length)
                .putInt(salt.length)
                .putInt(iv.length)
                .putInt(cipherText.length)
                .put(salt)
                .put(iv)
                .put(cipherText)
                .array();
    }

    private PasswordEncryptedData deserialize(byte[] serialized) {
        Objects.requireNonNull(serialized, "serialized");
        if (serialized.length < Integer.BYTES * 3) {
            throw new IllegalArgumentException("Encrypted payload is too short");
        }
        ByteBuffer buffer = ByteBuffer.wrap(serialized);
        int saltLength = readLength(buffer, "saltLength");
        int ivLength = readLength(buffer, "ivLength");
        int cipherTextLength = readLength(buffer, "cipherTextLength");
        int expectedLength = Integer.BYTES * 3 + saltLength + ivLength + cipherTextLength;
        if (expectedLength != serialized.length) {
            throw new IllegalArgumentException("Invalid encrypted payload length");
        }
        byte[] salt = new byte[saltLength];
        byte[] iv = new byte[ivLength];
        byte[] cipherText = new byte[cipherTextLength];
        buffer.get(salt);
        buffer.get(iv);
        buffer.get(cipherText);
        return new PasswordEncryptedData(salt, iv, cipherText);
    }

    private int readLength(ByteBuffer buffer, String name) {
        int length = buffer.getInt();
        if (length < 0) {
            throw new IllegalArgumentException(name + " must not be negative");
        }
        return length;
    }
}

File: modules/stego-core/src/main/kotlin/com/shterneregen/securelan/stego/service/impl/BmpSteganographyService.kt
package com.shterneregen.securelan.stego.service.impl

import com.shterneregen.securelan.crypto.model.PasswordEncryptedData
import com.shterneregen.securelan.crypto.workflow.PasswordFileCryptoWorkflow
import com.shterneregen.securelan.stego.model.BmpCapacity
import com.shterneregen.securelan.stego.model.ExtractedStegoPayload
import com.shterneregen.securelan.stego.model.StegoContentType
import com.shterneregen.securelan.stego.service.SteganographyService
import com.shterneregen.securelan.stego.service.impl.internal.BmpImage
import com.shterneregen.securelan.stego.service.impl.internal.StegoHeader
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.util.Objects

class BmpSteganographyService(
    private val passwordFileCryptoWorkflow: PasswordFileCryptoWorkflow,
) : SteganographyService {
    init {
        Objects.requireNonNull(passwordFileCryptoWorkflow, "passwordFileCryptoWorkflow")
    }

    override fun inspect(bmpBytes: ByteArray): BmpCapacity {
        val image = BmpImage.parse(bmpBytes)
        return BmpCapacity(
            image.width,
            image.height,
            image.bitsPerPixel,
            image.carrierByteCount,
            StegoHeader.BYTE_LENGTH,
            image.maxPayloadBytes(StegoHeader.BYTE_LENGTH),
        )
    }

    override fun hide(bmpBytes: ByteArray, payload: ByteArray, contentType: StegoContentType): ByteArray {
        Objects.requireNonNull(contentType, "contentType")
        return hideInternal(bmpBytes, payload, contentType, false)
    }

    override fun extract(bmpBytes: ByteArray): ExtractedStegoPayload {
        val image = BmpImage.parse(bmpBytes)
        val headerBytes = image.readLeastSignificantBits(bmpBytes, StegoHeader.BYTE_LENGTH)
        val header = StegoHeader.read(headerBytes)
        val maxPayloadBytes = image.maxPayloadBytes(StegoHeader.BYTE_LENGTH)
        if (header.payloadLength > maxPayloadBytes) {
            throw IllegalArgumentException("Hidden payload length exceeds BMP capacity")
        }
        val hiddenBytes = image.readLeastSignificantBits(bmpBytes, StegoHeader.BYTE_LENGTH + header.payloadLength)
        val payload = ByteArray(header.payloadLength)
        System.arraycopy(hiddenBytes, StegoHeader.BYTE_LENGTH, payload, 0, payload.size)
        return ExtractedStegoPayload(header.contentType, header.encrypted, payload)
    }

    override fun extractPayload(bmpBytes: ByteArray): ByteArray {
        val extracted = extract(bmpBytes)
        requirePlain(extracted)
        return extracted.payload()
    }

    override fun hideText(bmpBytes: ByteArray, message: String): ByteArray {
        Objects.requireNonNull(message, "message")
        return hide(bmpBytes, message.toByteArray(StandardCharsets.UTF_8), StegoContentType.UTF8_TEXT)
    }

    override fun extractText(bmpBytes: ByteArray): String {
        val extracted = extract(bmpBytes)
        requirePlain(extracted)
        return extracted.asUtf8String()
    }

    override fun hideEncryptedPayload(bmpBytes: ByteArray, payload: ByteArray, password: CharArray): ByteArray =
        hideEncrypted(bmpBytes, payload, StegoContentType.BINARY, password)

    override fun extractEncryptedPayload(bmpBytes: ByteArray, password: CharArray): ByteArray =
        extractEncrypted(bmpBytes, password, StegoContentType.BINARY)

    override fun hideEncryptedText(bmpBytes: ByteArray, message: String, password: CharArray): ByteArray {
        Objects.requireNonNull(message, "message")
        return hideEncrypted(bmpBytes, message.toByteArray(StandardCharsets.UTF_8), StegoContentType.UTF8_TEXT, password)
    }

    override fun extractEncryptedText(bmpBytes: ByteArray, password: CharArray): String =
        String(extractEncrypted(bmpBytes, password, StegoContentType.UTF8_TEXT), StandardCharsets.UTF_8)

    private fun hideEncrypted(bmpBytes: ByteArray, payload: ByteArray, contentType: StegoContentType, password: CharArray): ByteArray {
        Objects.requireNonNull(payload, "payload")
        requirePassword(password)
        val encryptedData = passwordFileCryptoWorkflow.encrypt(payload, password)
        return hideInternal(bmpBytes, serialize(encryptedData), contentType, true)
    }

    private fun extractEncrypted(bmpBytes: ByteArray, password: CharArray, expectedContentType: StegoContentType): ByteArray {
        requirePassword(password)
        val extracted = extract(bmpBytes)
        if (!extracted.encrypted()) {
            throw IllegalStateException("Extracted payload is not encrypted")
        }
        if (extracted.contentType() != expectedContentType) {
            throw IllegalStateException("Unexpected encrypted payload type: ${extracted.contentType()}")
        }
        val encryptedData = deserialize(extracted.payload())
        return passwordFileCryptoWorkflow.decrypt(encryptedData, password)
    }

    private fun hideInternal(bmpBytes: ByteArray, payload: ByteArray, contentType: StegoContentType, encrypted: Boolean): ByteArray {
        Objects.requireNonNull(payload, "payload")
        val image = BmpImage.parse(bmpBytes)
        val maxPayloadBytes = image.maxPayloadBytes(StegoHeader.BYTE_LENGTH)
        if (payload.size > maxPayloadBytes) {
            throw IllegalArgumentException("Payload is too large for this BMP: ${payload.size} bytes requested, $maxPayloadBytes bytes available")
        }
        val header = StegoHeader(contentType, encrypted, payload.size)
        val hiddenBytes = ByteBuffer.allocate(StegoHeader.BYTE_LENGTH + payload.size)
            .put(header.write())
            .put(payload)
            .array()
        val result = bmpBytes.clone()
        image.writeLeastSignificantBits(result, hiddenBytes)
        return result
    }

    private fun requirePlain(extracted: ExtractedStegoPayload) {
        if (extracted.encrypted()) {
            throw IllegalStateException("Extracted payload is encrypted; use encrypted extraction workflow")
        }
    }

    private fun requirePassword(password: CharArray) {
        Objects.requireNonNull(password, "password")
        if (password.isEmpty()) {
            throw IllegalArgumentException("password must not be empty")
        }
    }

    private fun serialize(encryptedData: PasswordEncryptedData): ByteArray {
        val salt = encryptedData.salt()
        val iv = encryptedData.iv()
        val cipherText = encryptedData.cipherText()
        return ByteBuffer.allocate(Int.SIZE_BYTES * 3 + salt.size + iv.size + cipherText.size)
            .putInt(salt.size)
            .putInt(iv.size)
            .putInt(cipherText.size)
            .put(salt)
            .put(iv)
            .put(cipherText)
            .array()
    }

    private fun deserialize(serialized: ByteArray): PasswordEncryptedData {
        Objects.requireNonNull(serialized, "serialized")
        if (serialized.size < Int.SIZE_BYTES * 3) {
            throw IllegalArgumentException("Encrypted payload is too short")
        }
        val buffer = ByteBuffer.wrap(serialized)
        val saltLength = readLength(buffer, "saltLength")
        val ivLength = readLength(buffer, "ivLength")
        val cipherTextLength = readLength(buffer, "cipherTextLength")
        val expectedLength = Int.SIZE_BYTES * 3 + saltLength + ivLength + cipherTextLength
        if (expectedLength != serialized.size) {
            throw IllegalArgumentException("Invalid encrypted payload length")
        }
        val salt = ByteArray(saltLength)
        val iv = ByteArray(ivLength)
        val cipherText = ByteArray(cipherTextLength)
        buffer.get(salt)
        buffer.get(iv)
        buffer.get(cipherText)
        return PasswordEncryptedData(salt, iv, cipherText)
    }

    private fun readLength(buffer: ByteBuffer, name: String): Int {
        val length = buffer.getInt()
        if (length < 0) {
            throw IllegalArgumentException("$name must not be negative")
        }
        return length
    }
}
